src/qcodes/parameters/infer.py

```diff
@@ -0,0 +1,144 @@
+from __future__ import annotations
+
+from collections.abc import Sequence
+from typing import TYPE_CHECKING, ClassVar
+
+from qcodes.instrument import Instrument, InstrumentBase, InstrumentModule
+from qcodes.instrument.parameter import DelegateParameter, Parameter
+
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+
+DOES_NOT_EXIST = "Does not exist"
+
+
+class InferError(AttributeError): ...
+
+
+class InferAttrs:
```

Comment:
Should we have any concerns for the fact that this will hold global state within a given python session? Is it really convenient and helpful to have this state or would it be opaque to users? And another concern about this state is - do we want to encourage a pattern where some attrs are added in the first place in the code, and then some further place in the code relies on the fact that the previous attrs were already added, rather than making it explicit which attrs it (this second place in the code) relies on?

Reply:
It feels to me like all state like this should live in the station and this should be a private attribute of the station that the functions can dive into

Reply:
What I was thinking here is that the number of non-Delegate linking parameters is likely to be small, but may be widely used. In particular, dyno Testpoints use `external_parameter` as the linking attribute. (for reasons I now regret), and the qcodes ScaledParameter uses `wrapped_parameter`. 

I do not want the experimenters to have to think about "Oh, does this use a dyno Testpoint or just a DelegateParameter or something else." I want inference to be completely transparent.

If we make the entries in the InferAttrs class Instance variables instead of class variables, then I need to start carrying around references to the specific instance in the Station, and I thought that felt clunky too. This seemed like a fairly harmless balance.

Reply:
ok, i see. Let's then live withthis for now and address this "attrbiutes state" and hopefully the "linking attributes" in the mid-/long-term.